# Telecom Churn Case Study Hackathon

# 1. Problem Statement

In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.

For many incumbent operators, retaining high profitable customers is the number one business
goal. To reduce customer churn, telecom companies need to predict which customers are at high risk of churn. In this project, you will analyze customer-level data of a leading telecom firm, build predictive models to identify customers at high risk of churn.

In this competition, your goal is to build a machine learning model that is able to predict churning customers based on the features provided for their usage.

# 2. Business Objectives

Predict churning customers for a Telecom company based on temporal behaviour

# 3. IMPORT LIBRARIES AND DATASETS

In [1]:
# import package
import numpy as np                                                 # Implemennts milti-dimensional array and matrices
import pandas as pd                                                # For data manipulation and analysis
import matplotlib.pyplot as plt                                    # Plotting library for Python programming language and it's numerical mathematics extension NumPy
import seaborn as sns                                              # Provides a high level interface for drawing attractive and informative statistical graphics

pd.set_option('display.max_columns', None)
%matplotlib inline
sns.set()

In [2]:
import warnings
warnings.filterwarnings('ignore')

## 3.1. Data Loading and Description

- **train.csv** - Contains 172 columns. The primary key that represents each customer is id. The target variable that you need to predict is churn_probability which contains a value of 0 or 1. This data is what you are going to use for EDA, cleaning, feature engineering, model building, model evaluation, model selection, and finally model training.
- **test.csv** - Contains 171 columns, doesn't contain the target variable churn_probability. You will apply all the necessary preprocessing steps to get this data into the right format and then use the model trained using the train.csv file to make predictions with this. This is unseen data! Only the competition hosts know the actual values of the target feature for this data and therefore your submissions will be evaluated on how well your model does with this dataset.
- **sample.csv** - This contains the format in which you need to submit the solutions to Kaggle. The id column in this dataset exactly the same as the id column in test.csv. You will make your predictions on the test.csv data and store them in a submission file that has the same format as this file. Check the Overview>Evaluation tab for mode details.
- **Data Dictionary** - This contains the definitions for the various acronyms that you will need to understand each variable. For example, the variable total_og_mou_7, contains the acronyms total, og, mou, and 7, which can be translated as the total outgoing minutes of voice calls made by the user in month of July.

To gain insights from data we must look into each aspect of it very carefully. We will start with observing few rows and columns of data both from the starting and from the end.

Let us check the basic information of the dataset. The very basic information to know is the dimension of the dataset – rows and columns – that’s what we find out with the method shape.

In [3]:
df = pd.read_csv("/kaggle/input/telecom-churn-case-study-hackathon-c42/train.csv")

In [4]:
df.sample(5)

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
15919  15919        109             0.0             0.0             0.0   
18508  18508        109             0.0             0.0             0.0   
49513  49513        109             0.0             0.0             0.0   
11510  11510        109             0.0             0.0             0.0   
10926  10926        109             0.0             0.0             0.0   

      last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
15919            6/30/2014            7/31/2014            8/31/2014   60.810   
18508            6/30/2014            7/31/2014            8/31/2014    9.319   
49513            6/30/2014            7/31/2014            8/31/2014  814.004   
11510            6/30/2014            7/31/2014            8/31/2014  733.934   
10926            6/30/2014            7/31/2014            8/31/2014    0.000   

        arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
15919   23.811   36.918        37.03        14.31        21.78         32.76   
18508  207.971    3.750         0.00        10.73         1.48         11.99   
49513  642.556  393.250       107.98        65.33        78.76       1019.09   
11510  615.580  378.580       177.29       118.46        54.04       1315.04   
10926   79.860  714.282          NaN        24.24        16.89           NaN   

       offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  \
15919          5.59          9.13            0.0            0.0   
18508        138.93          2.91            0.0            0.0   
49513        690.23        534.26            0.0            0.0   
11510       1160.74        674.71            0.0            0.0   
10926          8.01         35.31            NaN            0.0   

       roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
15919            0.0            0.0            0.0            0.0   
18508            0.0            0.0            0.0            0.0   
49513            0.0            0.0            0.0            0.0   
11510            0.0            0.0            0.0            0.0   
10926            0.0            NaN            0.0            0.0   

       loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
15919             37.03             14.31             21.78             32.76   
18508              0.00             10.73              1.48              5.11   
49513             40.66             27.76              8.73            218.43   
11510            124.19             94.76             43.68            246.96   
10926               NaN             24.24             16.89               NaN   

       loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
15919              5.59              9.13              0.00               0.0   
18508            113.04              2.91              0.00               0.0   
49513            214.59             38.99              0.48               0.0   
11510            382.64            265.78              0.00               0.0   
10926              6.96             22.43               NaN               0.0   

       loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
15919              0.00              0.00               0.0               0.0   
18508              0.00              4.26               0.0               0.0   
49513              0.00              0.00               0.0               0.0   
11510              0.88              0.00               0.0               0.0   
10926              0.00               NaN               0.0               0.0   

       loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
15919         69.79         19.91         30.91              0.00   
18508          5.11        123.78          4.39              0.00   
49513        259.58        242.36         47.73             67.31   
11510        371.16        477.41   

## 3.2. Data Understanding

In [5]:
df.shape

(69999, 172)

The dataset has __69999 rows and 172 columns.__

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


In [7]:
# Get info from both data set to understand data in depths 
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    id                        int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    arpu_6                    float64
 9    arpu_7                    float64
 10   arpu_8                    float64
 11   onnet_mou_6               float64
 12   onnet_mou_7               float64
 13   onnet_mou_8               float64
 14   offnet_mou_6              float64
 15   offnet_mou_7              float64
 16   offnet_mou_8              float64
 17   roam_ic_mou_6             float64
 18   roam_ic_mou_7             float64
 19   roam_ic_mou_8             float64
 20   roam

We have following 9 Columns as object type
- last_date_of_month_6
- last_date_of_month_7
- last_date_of_month_8
- date_of_last_rech_6
- date_of_last_rech_7
- date_of_last_rech_8
- date_of_last_rech_data_6
- date_of_last_rech_data_7
- date_of_last_rech_data_8

In [8]:
df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  69999.000000    69999.0         69297.0         69297.0   
mean   34999.000000      109.0             0.0             0.0   
std    20207.115084        0.0             0.0             0.0   
min        0.000000      109.0             0.0             0.0   
25%    17499.500000      109.0             0.0             0.0   
50%    34999.000000      109.0             0.0             0.0   
75%    52498.500000      109.0             0.0             0.0   
max    69998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         69297.0  69999.000000  69999.000000  69999.000000  67231.000000   
mean              0.0    283.134365    278.185912    278.858826    133.153275   
std               0.0    334.213918    344.366927    351.924315    299.963093   
min               0.0  -2258.709000  -1289.715000   -945.808000      0.000000   
25%               0.0     93.581000     86.714000     84.095000      7.410000   
50%               0.0    197.484000    191.588000    192.234000     34.110000   
75%               0.0    370.791000    365.369500    369.909000    119.390000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  67312.000000  66296.000000  67231.000000  67312.000000  66296.000000   
mean     133.894438    132.978257    198.874771    197.153383    196.543577   
std      311.277193    311.896596    316.818355    322.482226    324.089234   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.675000      6.410000     34.860000     32.240000     31.575000   
50%       32.280000     32.100000     96.480000     91.885000     91.800000   
75%      115.837500    115.060000    232.990000    227.630000    229.345000   
max     8157.780000  10752.560000   8362.360000   7043.980000  14007.340000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   67231.000000   67312.000000   66296.000000   67231.000000   
mean        9.765435       7.014568       7.004892      14.186457   
std        57.374429      55.960985      53.408135      73.469261   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max      2850.980000    4155.830000    4169.810000    3775.110000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   67312.000000   66296.000000      67231.000000      67312.000000   
mean        9.842191       9.771783         46.904854         46.166503   
std        58.511894      64.618388        150.971758        154.739002   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.650000   
50%         0.000000       0.000000         11.910000         11.580000   
75%         0.000000       0.000000         40.740000         39.760000   
max      2812.040000    5337.040000       6431.330000       7400.660000   

       loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
count      66296.000000      67231.000000      67312.000000      66296.000000   
mean          45.686109         93.238231         90.799240         91.121447   
std          153.716880        162.046699        153.852597        152.997805   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.610000          9.920000         10.090000          9.830000   
50%           11.740000         41.030000         40.170000         40.350000   
75%           39.895000        110.430000        107.540000        109.245000   
max        10752.560000       4696

- The dataset comprises of 69999 observations of 172 columns. Above is a table showing names of all the columns and their description.
- There are no categorical value present in dataset, rest of columns are integer/float type.
- Also few columns have object type as date value which we can use to create a derive data from that

# 4. Data Engineering 

Below are the points that need to be addressed in this step:
1. Duplicates
2. Data validity check
3. Missing values
4. Outliers

## 4.1. Data Duplicates

The data may contain duplicate entries, which need to be removed in most of the applications.

In [9]:
duplicate_df = df[df.duplicated()]
duplicate_df.shape

(0, 172)

There are duplicates entrires present in traing dataset

## 4.2. Data validity check

Based on the features, need to validate the data.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


We have arount 172 features. However there are 9 Columns as object type & might not useful for our analysis.

- last_date_of_month_6
- last_date_of_month_7
- last_date_of_month_8
- date_of_last_rech_6
- date_of_last_rech_7
- date_of_last_rech_8
- date_of_last_rech_data_6
- date_of_last_rech_data_7
- date_of_last_rech_data_8

## 4.3. Missing values

- Based on the business, a missing value imputation needs to be chosen.
- If the data is normally distributed, mean imputation can be performed. If the mean and median difference is huge, median imputation is preferred.

In [11]:
# Checking more meta information about dataframe i.e. how many non-null, uniquie value.

def metadataInfo(dataframe) : 
    return pd.DataFrame({
                'Datatype' : dataframe.dtypes.astype(str), 
                'Non_Null_Count': dataframe.count(axis = 0).astype(int), 
                'Null_Count': dataframe.isnull().sum().astype(int), 
                'Null_Percentage': round(dataframe.isnull().sum()/len(dataframe) * 100 , 2), 
                'Unique_Values_Count': dataframe.nunique().astype(int) 
                 }).sort_values(by='Null_Percentage', ascending=False)

In [12]:
new_df = metadataInfo(df)
print(new_df.to_string())

                         Datatype  Non_Null_Count  Null_Count  Null_Percentage  Unique_Values_Count
arpu_3g_6                 float64           17568       52431            74.90                 5506
count_rech_2g_6           float64           17568       52431            74.90                   30
night_pck_user_6          float64           17568       52431            74.90                    2
arpu_2g_6                 float64           17568       52431            74.90                 5389
date_of_last_rech_data_6   object           17568       52431            74.90                   30
total_rech_data_6         float64           17568       52431            74.90                   36
av_rech_amt_data_6        float64           17568       52431            74.90                  792
max_rech_data_6           float64           17568       52431            74.90                   47
count_rech_3g_6           float64           17568       52431            74.90                   23


There are so many __Null/NA__ value present in dataset which varies from 1% to 74%.
- Columns which have more that 60% __NULL/NA__ value, we will drop those since we might not more information from those colum
- However columns which have minimum Null value present and if the data is normally distributed in those, then mean imputation can be performed. If the mean and median difference is huge, median imputation is preferred on numerical data.
- Since we don't have categorical data present, no need to perform Mode imputation on categorical data.

In [13]:
# some recharge columns have minimum value of 1 while some don't have
recharge_cols = ['total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8', 
                 'count_rech_2g_6', 'count_rech_2g_7', 'count_rech_2g_8', 
                 'count_rech_3g_6', 'count_rech_3g_7', 'count_rech_3g_8', 
                 'max_rech_data_6', 'max_rech_data_7', 'max_rech_data_8', 
                 'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8', 
                 ]

df[recharge_cols].describe(include='all')

total_rech_data_6  total_rech_data_7  total_rech_data_8  \
count       17568.000000       17865.000000       18417.000000   
mean            2.467612           2.679989           2.652441   
std             2.794610           3.073472           3.101265   
min             1.000000           1.000000           1.000000   
25%             1.000000           1.000000           1.000000   
50%             1.000000           2.000000           1.000000   
75%             3.000000           3.000000           3.000000   
max            61.000000          54.000000          60.000000   

       count_rech_2g_6  count_rech_2g_7  count_rech_2g_8  count_rech_3g_6  \
count     17568.000000     17865.000000     18417.000000     17568.000000   
mean          1.865323         2.056311         2.016018         0.602288   
std           2.566377         2.799916         2.728246         1.279297   
min           0.000000         0.000000         0.000000         0.000000   
25%           1.000000         1.000000         1.000000         0.000000   
50%           1.000000         1.000000         1.000000         0.000000   
75%           2.000000         2.000000         2.000000         1.000000   
max          42.000000        48.000000        44.000000        29.000000   

       count_rech_3g_7  count_rech_3g_8  max_rech_data_6  max_rech_data_7  \
count     17865.000000     18417.000000     17568.000000     17865.000000   
mean          0.623678         0.636423       126.500000       126.402071   
std           1.401230         1.457058       109.352573       109.459266   
min           0.000000         0.000000         1.000000         1.000000   
25%           0.000000         0.000000        25.000000        25.000000   
50%           0.000000         0.000000       145.000000       145.000000   
75%           1.000000         1.000000       177.000000       177.000000   
max          34.000000        45.000000      1555.000000      1555.000000   

       max_rech_data_8  av_rech_amt_data_6  av_rech_amt_data_7  \
count     18417.000000        17568.000000        17865.000000   
mean        125.374925          192.831096          201.455940   
std         109.648799          190.623115          198.346141   
min           1.000000            1.000000            1.000000   
25%          25.000000           82.000000           92.000000   
50%         145.000000          154.000000          154.000000   
75%         179.000000          252.000000          252.000000   
max        1555.000000         5920.000000         4365.000000   

       av_rech_amt_data_8  
count        18417.000000  
mean           196.815792  
std            192.280532  
min              1.000000  
25%             84.000000  
50%            154.000000  
75%            252.000000  
max           4076.000000

- We can create new feature as __total_rech_amt_data__ using __total_rech_data__ and __av_rech_amt_data__ to capture amount utilized by customer for data.
- Also as the minimum value is 1 we can impute the NA values by 0, Considering there were no recharges done by the customer.

In [14]:
# It is also observed that the recharge date and the recharge value are missing together which means the customer didn't recharge
df.loc[df.total_rech_data_6.isnull() & df.date_of_last_rech_data_6.isnull(), ["total_rech_data_6", "date_of_last_rech_data_6"]].head(5)

total_rech_data_6 date_of_last_rech_data_6
0                NaN                      NaN
1                NaN                      NaN
2                NaN                      NaN
3                NaN                      NaN
5                NaN                      NaN

In the recharge variables where minumum value is 1, we can impute missing values with zeroes since it means customer didn't recharge their numbers that month.

In [15]:
# create a list of recharge columns where we will impute missing values with zeroes
zero_impute = ['total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8', 'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8', 
               'max_rech_data_6', 'max_rech_data_7', 'max_rech_data_8']

In [16]:
# impute missing values with 0
df[zero_impute] = df[zero_impute].apply(lambda x: x.fillna(0))

In [17]:
# now we have to  make sure the values are imputed correctly for that we can check "Missing value ratio"
df[zero_impute].isnull().sum()*100/df.shape[1]

total_rech_data_6     0.0
total_rech_data_7     0.0
total_rech_data_8     0.0
av_rech_amt_data_6    0.0
av_rech_amt_data_7    0.0
av_rech_amt_data_8    0.0
max_rech_data_6       0.0
max_rech_data_7       0.0
max_rech_data_8       0.0
dtype: float64

In [18]:
# now we can check the "statistics Summary"
df[zero_impute].describe(include='all')

total_rech_data_6  total_rech_data_7  total_rech_data_8  \
count       69999.000000       69999.000000       69999.000000   
mean            0.619309           0.683981           0.697867   
std             1.762010           1.943193           1.973434   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           1.000000           1.000000   
max            61.000000          54.000000          60.000000   

       av_rech_amt_data_6  av_rech_amt_data_7  av_rech_amt_data_8  \
count        69999.000000        69999.000000        69999.000000   
mean            48.395787           51.415168           51.782975   
std            126.923180          133.246541          131.291162   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              8.000000           17.000000           23.000000   
max           5920.000000         4365.000000         4076.000000   

       max_rech_data_6  max_rech_data_7  max_rech_data_8  
count     69999.000000     69999.000000     69999.000000  
mean         31.748339        32.260075        32.986614  
std          77.519536        78.069163        78.808337  
min           0.000000         0.000000         0.000000  
25%           0.000000         0.000000         0.000000  
50%           0.000000         0.000000         0.000000  
75%           8.000000        14.000000        17.000000  
max        1555.000000      1555.000000      1555.000000

In [19]:
# now we can create some column name list by there types using description of columns
id_cols = ['id', 'circle_id']

date_cols = ['last_date_of_month_6',
             'last_date_of_month_7',
             'last_date_of_month_8',             
             'date_of_last_rech_6',
             'date_of_last_rech_7',
             'date_of_last_rech_8',             
             'date_of_last_rech_data_6',
             'date_of_last_rech_data_7',
             'date_of_last_rech_data_8'             
            ]

cat_cols =  ['night_pck_user_6',
             'night_pck_user_7',
             'night_pck_user_8',             
             'fb_user_6',
             'fb_user_7',
             'fb_user_8'             
            ]

num_cols = [column for column in df.columns if column not in id_cols + date_cols + cat_cols]

# print the number of columns in each list
print("#ID cols: %d\n#Date cols:%d\n#Numeric cols:%d\n#Category cols:%d" % (len(id_cols), len(date_cols), len(num_cols), len(cat_cols)))

# check if we have missed any column or not
print(len(id_cols) + len(date_cols) + len(num_cols) + len(cat_cols) == df.shape[1])

#ID cols: 2
#Date cols:9
#Numeric cols:155
#Category cols:6
True


In [20]:
# drop id and date columns
df = df.drop(id_cols + date_cols, axis=1)
#check the shape again
df.shape

(69999, 161)

In [21]:
# replace missing values with '-1' in categorical columns
df[cat_cols] = df[cat_cols].apply(lambda x: x.fillna(-1))

In [22]:
# missing value ratio
df[cat_cols].isnull().sum()*100/df.shape[0]

night_pck_user_6    0.0
night_pck_user_7    0.0
night_pck_user_8    0.0
fb_user_6           0.0
fb_user_7           0.0
fb_user_8           0.0
dtype: float64

In [23]:
# Droping variables with more than 70% of missing values (we can call it as threshold )

initial_cols = df.shape[1]

MISSING_THRESHOLD = 0.7

include_cols = list(df.apply(lambda column: True if column.isnull().sum()/df.shape[0] < MISSING_THRESHOLD else False))

drop_missing = pd.DataFrame({'features':df.columns , 'include': include_cols})
drop_missing.loc[drop_missing.include == True,:]

features  include
0       loc_og_t2o_mou     True
1       std_og_t2o_mou     True
2       loc_ic_t2o_mou     True
3               arpu_6     True
4               arpu_7     True
..                 ...      ...
156                aon     True
157         aug_vbc_3g     True
158         jul_vbc_3g     True
159         jun_vbc_3g     True
160  churn_probability     True

[149 rows x 2 columns]

In [24]:
# now we can drop  some more columns
df = df.loc[:, include_cols]

dropped_cols = df.shape[1] - initial_cols
dropped_cols

-12

In [25]:
#rechecking the shape of a dataframe
df.shape

(69999, 149)

In [26]:
# Let's have a look of the missing values for how many missing values has left
print(metadataInfo(df).to_string())

                   Datatype  Non_Null_Count  Null_Count  Null_Percentage  Unique_Values_Count
std_ic_t2t_mou_8    float64           66296        3703             5.29                 5367
spl_og_mou_8        float64           66296        3703             5.29                 3848
loc_og_t2c_mou_8    float64           66296        3703             5.29                 2204
loc_og_mou_8        float64           66296        3703             5.29                21950
std_og_t2t_mou_8    float64           66296        3703             5.29                14483
std_og_t2m_mou_8    float64           66296        3703             5.29                15968
std_og_t2f_mou_8    float64           66296        3703             5.29                 1945
std_og_t2c_mou_8    float64           66296        3703             5.29                    1
std_og_mou_8        float64           66296        3703             5.29                21745
isd_og_mou_8        float64           66296        3703     

In [27]:
num_cols = [column for column in df.columns if column not in id_cols + date_cols + cat_cols]
num_cols

['loc_og_t2o_mou',
 'std_og_t2o_mou',
 'loc_ic_t2o_mou',
 'arpu_6',
 'arpu_7',
 'arpu_8',
 'onnet_mou_6',
 'onnet_mou_7',
 'onnet_mou_8',
 'offnet_mou_6',
 'offnet_mou_7',
 'offnet_mou_8',
 'roam_ic_mou_6',
 'roam_ic_mou_7',
 'roam_ic_mou_8',
 'roam_og_mou_6',
 'roam_og_mou_7',
 'roam_og_mou_8',
 'loc_og_t2t_mou_6',
 'loc_og_t2t_mou_7',
 'loc_og_t2t_mou_8',
 'loc_og_t2m_mou_6',
 'loc_og_t2m_mou_7',
 'loc_og_t2m_mou_8',
 'loc_og_t2f_mou_6',
 'loc_og_t2f_mou_7',
 'loc_og_t2f_mou_8',
 'loc_og_t2c_mou_6',
 'loc_og_t2c_mou_7',
 'loc_og_t2c_mou_8',
 'loc_og_mou_6',
 'loc_og_mou_7',
 'loc_og_mou_8',
 'std_og_t2t_mou_6',
 'std_og_t2t_mou_7',
 'std_og_t2t_mou_8',
 'std_og_t2m_mou_6',
 'std_og_t2m_mou_7',
 'std_og_t2m_mou_8',
 'std_og_t2f_mou_6',
 'std_og_t2f_mou_7',
 'std_og_t2f_mou_8',
 'std_og_t2c_mou_6',
 'std_og_t2c_mou_7',
 'std_og_t2c_mou_8',
 'std_og_mou_6',
 'std_og_mou_7',
 'std_og_mou_8',
 'isd_og_mou_6',
 'isd_og_mou_7',
 'isd_og_mou_8',
 'spl_og_mou_6',
 'spl_og_mou_7',
 'spl_og_mou

In [28]:
#imputing with meadian for num_cols
df[num_cols] = df[num_cols].apply(lambda x: x.fillna(x.median()))

In [29]:
print(metadataInfo(df).to_string())

                   Datatype  Non_Null_Count  Null_Count  Null_Percentage  Unique_Values_Count
loc_og_t2o_mou      float64           69999           0              0.0                    1
isd_ic_mou_7        float64           69999           0              0.0                 4640
ic_others_6         float64           69999           0              0.0                 1523
ic_others_7         float64           69999           0              0.0                 1658
ic_others_8         float64           69999           0              0.0                 1565
total_rech_num_6      int64           69999           0              0.0                   94
total_rech_num_7      int64           69999           0              0.0                   96
total_rech_num_8      int64           69999           0              0.0                   92
total_rech_amt_6      int64           69999           0              0.0                 2108
total_rech_amt_7      int64           69999           0     

Now we don't any null value present in dataset

## 4.4. Outliers

- Outliers can be valid outlier or data entry mistake. 
- Need to understand what kind of outlier the data contains. Based on the outlier type, the issue needs to be addressed.

# 5. Exploratory Data Analysis 

This step gives more insights into the data and how the data is related to the target variable. This step involves mainly:
1. Uni-variate analysis
2. Bi-variate/Multi-variate analysis
3. Pivots
4. Visualization and Data Insights

# 6. Feature Engineering and Selection

- Feature engineering involves identifying the right drivers/features that affect the target variable and also deriving the new features based on existing features.
- Based on feature importance, some of the features can be removed from the data, which helps in reducing data size. 

# 7. Model Building

Model building is the process of developing a probabilistic model that best describes the relationship between independent and dependent variables. Various ML models are to be built based on the problem statement.

Below are the topics that can be covered in this step :
1. Creating dummy variables
2. Over Sampling and Under Sampling (if the data is imbalanced)
3. Split the data into train and test

# 8. Model Validation & Evaluation

This step covers
1. Testing the model
2. Tuning the model
3. Cross-validation
4. Model evaluation metrics trade-off
5. Model Underfitting/Overfitting

## 8.1. Testing the model :

Run the model on test data and evaluate the model performance using the correct metric based on the business use case.

## 8.2. Tuning the model :

Model tuning involves, improving model performance by iterating the parameter values during model building. After fine-tuning, the model needs to re-build.

## 8.3. Cross-validation :

Cross-validation is used to evaluate how the model will perform on an independent test dataset. Some of the cross-validation techniques are :
1. K- fold cross-validation

## 8.4. Model Evaluation metrics trade-off :

Trade-offs always help us to find the sweet spot or the middle ground. Machine learning mostly deals with two tradeoffs :
1. Bias-variance tradeoff
2. Precision-Recall tradeoff

## 8.5. Model Underfitting and Overfitting :

- Overfitting is the case, where the model is trying to capture all the patterns in the training data and failing to perform in the test data.
- Underfitting is the case, where the model is not learning properly in the training data and also not performing well in test data.

# 9. Model Predict

# 10. Conclusion 